File: modules/apps/desktop/flashcards/src/today.test.ts
/**
 * The day the window weighs a goal against is the review day, which begins at
 * the hour the settings name. Between midnight and that hour the calendar has
 * moved on and the review day has not, and a preset due today would be called
 * over while the core still schedules it.
 */
import { afterEach, describe, expect, it, vi } from 'vitest'
import { Goal, Stopped } from '@numen/protocol'

import { counting } from './counting'
import type { Counts } from './counting'
import { named, scheduling } from './scheduling'
import type { Asks, SettingsMessage } from './scheduling'
import type { Owing } from './core'

const dated = (day: string): SettingsMessage => ({
  goal: Goal.BY_DATE,
  byDate: day,
  minutesADay: 20,
  newADay: 10,
  reviewsADay: 200,
  retention: 0.9,
  load: {},
  evenLoad: true,
})

const vault: Owing = {
  vaultId: '01A',
  name: 'Vault',
  path: '/vaults/01A',
  counted: true,
  faces: 4,
  due: 3,
  new: 1,
  decks: [{ deck: 'decks/Words.md', faces: 4, due: 3, new: 1, learned: 2, unbegun: 1 }],
  presets: [],
  unread: '',
  reading: false,
}

const answering = (settings: SettingsMessage): Asks => ({
  async scheduling() {
    return {
      preset: {
        path: 'Sanskrit.md',
        title: 'Sanskrit',
        settings,
        problems: [],
        stopsOn: Stopped.NOTHING,
      },
    }
  },
})

afterEach(() => {
  vi.useRealTimers()
})

describe('the day a goal is weighed against', () => {
  it('is the one the application counted', async () => {
    const cards: Counts = {
      async *owing() {
        yield { day: '2026-09-04', vaults: [] }
      },
    }
    const held = counting({ cards, failed: () => {} })

    await held.count()

    expect(held.day.value).toBe('2026-09-04')
  })

  it('leaves a preset due today running in the hours past midnight', async () => {
    // One in the morning: the calendar says the fifth, and the review day that
    // began at four on the fourth is still running.
    vi.useFakeTimers()
    vi.setSystemTime(new Date(2026, 8, 5, 1, 0, 0))
    expect(named(new Date())).toBe('2026-09-05')

    const cards: Counts = {
      async *owing() {
        yield { day: '2026-09-04', vaults: [] }
      },
    }
    const held = counting({ cards, failed: () => {} })
    await held.count()

    const one = scheduling({ presets: answering(dated('2026-09-04')) })
    await one.read(vault, held.day.value)

    expect(one.presets.value[0]?.paused).toBe('')
    expect(one.presets.value[0]?.cards).toBe(4)
  })
})

File: modules/libs/core/adapter/flashcardsui/today_test.go
package flashcardsui

import (
	"testing"
	"time"

	"github.com/jiva-studio/numen/modules/libs/core/review"
)

// The counts the window is handed stand in a review day, and the window weighs
// a preset's goal against that day. A day of review begins at the hour the
// settings name, so between midnight and that hour the calendar has turned and
// the day has not.
func TestTheDayTheCountsStandInIsTheReviewDay(t *testing.T) {
	api, _ := windowed(t, deck)
	api.Day = review.Day{Starts: 4 * time.Hour, In: time.UTC}

	for name, c := range map[string]struct {
		at   time.Time
		want string
	}{
		"an hour past midnight": {
			time.Date(2026, 9, 5, 1, 0, 0, 0, time.UTC), "2026-09-04",
		},
		"the hour the day begins at": {
			time.Date(2026, 9, 5, 4, 0, 0, 0, time.UTC), "2026-09-05",
		},
		"the middle of the day": {
			time.Date(2026, 9, 5, 14, 0, 0, 0, time.UTC), "2026-09-05",
		},
	} {
		t.Run(name, func(t *testing.T) {
			api.Now = func() time.Time { return c.at }
			if got := front(t, api).GetDay(); got != c.want {
				t.Errorf("at %v the counts stand in %q, want %q", c.at, got, c.want)
			}
		})
	}
}
